File: src/models/token.rs
use chrono::{DateTime, NaiveDateTime, Utc};
use serde::{Deserialize, Serialize};

pub const API_TOKEN_NAME_MAX_CHARS: usize = 80;
pub const API_TOKEN_DESCRIPTION_MAX_CHARS: usize = 200;
/// A custom token is no longer floored at 16 bytes. The console warns and asks
/// for a second confirmation below that, and the threshold lives there
/// (TOKEN_WEAK_BYTES in static/js/tokens.js) rather than here: a stateless
/// request cannot distinguish an operator who was warned and accepted from one
/// who never saw the warning, so enforcing it server-side would either refuse
/// the confirmed case or need a "yes I mean it" flag that any client could set.
/// What is left here is what is structurally invalid — empty, or too long.
pub const API_TOKEN_MIN_BYTES: usize = 1;
pub const API_TOKEN_MAX_BYTES: usize = 256;

/// The shape SQLite's `datetime('now')` produces, and the only shape
/// `expires_at` is ever stored in. Fixed width and zero padded, so lexical
/// order is chronological order — which is what lets the authentication SQL in
/// `middleware::auth` and the checks in `db::token` compare expiries as plain
/// strings and always reach the same verdict.
pub const TIMESTAMP_FORMAT: &str = "%Y-%m-%d %H:%M:%S";

const EXPIRY_FORMAT_ERROR: &str =
    "token expiry must be an RFC 3339 timestamp or 'YYYY-MM-DD HH:MM:SS' in UTC";

/// Now, in the stored timestamp shape, for comparison against an expiry.
pub fn utc_now_timestamp() -> String {
    Utc::now().format(TIMESTAMP_FORMAT).to_string()
}

/// Normalize a caller-supplied expiry into the stored UTC shape.
///
/// A blank value means "never expires" and is reported as `None`, so clearing
/// the console's expiry field behaves the same whether the client sends `null`
/// or `""`. Whether the result lies in the past is not decided here — that
/// depends on the row being written, and lives in `db::token`.
pub fn normalize_expires_at(raw: Option<&str>) -> Result<Option<String>, &'static str> {
    let Some(value) = raw.map(str::trim).filter(|value| !value.is_empty()) else {
        return Ok(None);
    };
    let utc = if let Ok(parsed) = DateTime::parse_from_rfc3339(value) {
        parsed.with_timezone(&Utc)
    } else if let Ok(naive) = NaiveDateTime::parse_from_str(value, TIMESTAMP_FORMAT) {
        naive.and_utc()
    } else {
        return Err(EXPIRY_FORMAT_ERROR);
    };
    Ok(Some(utc.format(TIMESTAMP_FORMAT).to_string()))
}

// ── DB row ──────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, sqlx::FromRow)]
pub struct ApiTokenRow {
    pub id: i64,
    pub name: String,
    pub description: String,
    pub token_preview: String,
    pub enabled: i64, // 0 / 1
    pub expires_at: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

// ── Input models ─────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ApiTokenIn {
    pub name: String,
    #[serde(default)]
    pub description: String,
    /// If None, a token will be auto-generated.
    pub token: Option<String>,
    #[serde(default = "default_enabled")]
    pub enabled: bool,
    /// Absent, null or blank means the token never expires.
    #[serde(default)]
    pub expires_at: Option<String>,
}

/// The update endpoint is a full replacement, so an absent `expires_at` clears
/// the expiry rather than leaving it alone. The console always sends the field.
#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ApiTokenUpdateIn {
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub expires_at: Option<String>,
}

fn default_enabled() -> bool {
    true
}

fn validate_metadata(name: &str, description: &str) -> Result<(), &'static str> {
    if name.trim().is_empty() || name.chars().count() > API_TOKEN_NAME_MAX_CHARS {
        return Err("token name must be between 1 and 80 characters");
    }
    if name.chars().any(char::is_control) {
        return Err("token name must not contain control characters");
    }
    if description.chars().count() > API_TOKEN_DESCRIPTION_MAX_CHARS {
        return Err("token description must be at most 200 characters");
    }
    if description.chars().any(char::is_control) {
        return Err("token description must not contain control characters");
    }
    Ok(())
}

impl ApiTokenIn {
    pub fn validate(&self) -> Result<(), &'static str> {
        validate_metadata(&self.name, &self.description)?;
        self.normalized_expires_at()?;
        let Some(token) = self.token.as_deref() else {
            return Ok(());
        };
        if !(API_TOKEN_MIN_BYTES..=API_TOKEN_MAX_BYTES).contains(&token.len()) {
            return Err("custom token must be between 1 and 256 bytes");
        }
        if !token.bytes().all(|byte| byte.is_ascii_graphic()) {
            return Err("custom token must contain only printable ASCII characters without spaces");
        }
        Ok(())
    }

    pub fn normalized_expires_at(&self) -> Result<Option<String>, &'static str> {
        normalize_expires_at(self.expires_at.as_deref())
    }
}

impl ApiTokenUpdateIn {
    pub fn validate(&self) -> Result<(), &'static str> {
        validate_metadata(&self.name, &self.description)?;
        self.normalized_expires_at()?;
        Ok(())
    }

    pub fn normalized_expires_at(&self) -> Result<Option<String>, &'static str> {
        normalize_expires_at(self.expires_at.as_deref())
    }
}

// ── Output models ───────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize)]
pub struct ApiTokenOut {
    pub id: i64,
    pub name: String,
    pub description: String,
    pub token_preview: String,
    pub enabled: bool,
    pub expires_at: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

/// Returned only by the creation endpoint so the full token can be shown once.
#[derive(Debug, Clone, Serialize)]
pub struct ApiTokenCreatedOut {
    pub id: i64,
    pub name: String,
    pub description: String,
    pub token: String,
    pub token_preview: String,
    pub enabled: bool,
    pub expires_at: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[cfg(test)]
mod tests {
    use super::{normalize_expires_at, ApiTokenIn, ApiTokenOut, ApiTokenUpdateIn};

    fn valid_create_input() -> ApiTokenIn {
        ApiTokenIn {
            name: "client A".into(),
            description: "integration token".into(),
            token: Some("custom-token-value-1234".into()),
            enabled: true,
            expires_at: None,
        }
    }

    #[test]
    fn validates_custom_token_and_metadata_boundaries() {
        assert!(valid_create_input().validate().is_ok());

        for token in ["", "contains whitespace", &"x".repeat(257)] {
            let mut input = valid_create_input();
            input.token = Some(token.into());
            assert!(input.validate().is_err(), "{token:?} must be rejected");
        }

        // Short custom tokens are accepted here on purpose. The console warns
        // and asks for a second confirmation below 16 bytes; the server cannot
        // tell a warned-and-accepted request from an unwarned one, so it only
        // rejects what is structurally invalid.
        for token in ["x", "too-short"] {
            let mut input = valid_create_input();
            input.token = Some(token.into());
            assert!(input.validate().is_ok(), "{token:?} must be accepted");
        }

        let mut input = valid_create_input();
        input.name = "   ".into();
        assert!(input.validate().is_err());

        let mut input = valid_create_input();
        input.description = "x".repeat(201);
        assert!(input.validate().is_err());

        let update = ApiTokenUpdateIn {
            name: "valid".into(),
            description: "line\nbreak".into(),
            expires_at: None,
        };
        assert!(update.validate().is_err());
    }

    #[test]
    fn expiry_normalizes_every_accepted_form_to_stored_utc() {
        // Absent, null and blank all mean the same thing: never expires.
        for raw in [None, Some(""), Some("   ")] {
            assert_eq!(normalize_expires_at(raw), Ok(None));
        }

        // An offset is honored, then converted; a bare timestamp is UTC already.
        for raw in [
            "2026-09-01T12:00:00Z",
            "2026-09-01T20:00:00+08:00",
            "2026-09-01 12:00:00",
        ] {
            assert_eq!(
                normalize_expires_at(Some(raw)),
                Ok(Some("2026-09-01 12:00:00".into())),
                "{raw} must normalize to stored UTC"
            );
        }

        for raw in [
            "2026-09-01",          // no time of day
            "2026-09-01 12:00",    // no seconds
            "01/09/2026 12:00:00", // not ISO order
            "30d",                 // relative expressions are the console's job
            "tomorrow",
        ] {
            assert!(
                normalize_expires_at(Some(raw)).is_err(),
                "{raw} must be rejected"
            );
        }
    }

    #[test]
    fn a_malformed_expiry_fails_validation_on_both_endpoints() {
        let mut input = valid_create_input();
        input.expires_at = Some("not a timestamp".into());
        assert!(input.validate().is_err());

        input.expires_at = Some("2026-09-01T12:00:00Z".into());
        assert!(input.validate().is_ok());

        let update = ApiTokenUpdateIn {
            name: "valid".into(),
            description: String::new(),
            expires_at: Some("not a timestamp".into()),
        };
        assert!(update.validate().is_err());
    }

    #[test]
    fn regular_output_never_serializes_a_full_token() {
        let value = serde_json::to_value(ApiTokenOut {
            id: 1,
            name: "client".into(),
            description: String::new(),
            token_preview: "wildtoke…".into(),
            enabled: true,
            expires_at: Some("2026-09-01 12:00:00".into()),
            created_at: "2026-01-01 00:00:00".into(),
            updated_at: "2026-01-01 00:00:00".into(),
        })
        .unwrap();

        assert!(value.get("token").is_none());
        assert_eq!(value["token_preview"], "wildtoke…");
    }
}

File: internal/models/token.go
package models

import (
	"strings"
	"time"
	"unicode"
	"unicode/utf8"

	"github.com/liguangsheng/wildtoken/internal/ratelimit"
)

const (
	APITokenNameMaxChars        = 80
	APITokenDescriptionMaxChars = 200
	// APITokenMinBytes no longer floors a custom token at 16 bytes. The console
	// warns and asks for a second confirmation below that, and the threshold
	// lives there (TOKEN_WEAK_BYTES in static/js/tokens.js) rather than here: a
	// stateless request cannot distinguish an operator who was warned and
	// accepted from one who never saw the warning, so enforcing it server-side
	// would either refuse the confirmed case or need a "yes I mean it" flag that
	// any client could set. What is left here is what is structurally invalid —
	// empty, or too long.
	APITokenMinBytes = 1
	APITokenMaxBytes = 256
)

// TimestampFormat is the shape SQLite's `datetime('now')` produces, and the only
// shape `expires_at` is ever stored in. Fixed width and zero padded, so lexical
// order is chronological order — which is what lets the authentication SQL in
// middleware and the checks in the token store compare expiries as plain strings
// and always reach the same verdict.
const TimestampFormat = "2006-01-02 15:04:05"

const expiryFormatError = ErrString(
	"token expiry must be an RFC 3339 timestamp or 'YYYY-MM-DD HH:MM:SS' in UTC")

// UTCNowTimestamp renders now in the stored timestamp shape, for comparison
// against an expiry.
func UTCNowTimestamp() string {
	return time.Now().UTC().Format(TimestampFormat)
}

// NormalizeExpiresAt converts a caller-supplied expiry into the stored UTC shape.
//
// A blank value means "never expires" and is reported as nil, so clearing the
// console's expiry field behaves the same whether the client sends `null` or
// `""`. Whether the result lies in the past is not decided here — that depends
// on the row being written, and lives in the token store.
func NormalizeExpiresAt(raw *string) (*string, error) {
	if raw == nil {
		return nil, nil
	}
	value := strings.TrimSpace(*raw)
	if value == "" {
		return nil, nil
	}
	if parsed, err := time.Parse(time.RFC3339, value); err == nil {
		normalized := parsed.UTC().Format(TimestampFormat)
		return &normalized, nil
	}
	if parsed, err := time.Parse(TimestampFormat, value); err == nil {
		normalized := parsed.UTC().Format(TimestampFormat)
		return &normalized, nil
	}
	return nil, expiryFormatError
}

// APITokenRow mirrors a row of the `api_tokens` table.
type APITokenRow struct {
	ID           int64
	Name         string
	Description  string
	TokenPreview string
	// Token is the stored plaintext, empty for rows written before it was kept.
	Token       string
	Enabled     int64
	ExpiresAt   *string
	CreatedAt   string
	UpdatedAt   string
	GroupID     int64
	UsedTokens  int64
	LimitTokens *int64
	RateLimit   *string
	// AllowedModels is the stored JSON array. Nil covers a row written before the
	// column existed; both nil and "[]" mean unrestricted.
	AllowedModels *string
	// QuotaPeriod controls automatic quota rollover. None is the legacy lifetime
	// total; the other values bind usage to a calendar period in QuotaTimezone.
	QuotaPeriod   string
	QuotaTimezone string
	// QuotaPeriodStamp is the cycle UsedTokens was accumulated under, which is not
	// necessarily the current one — the counter is cleared by the first usage of a
	// new period, not on the boundary itself.
	QuotaPeriodStamp string
}

// UsedTokensNow reports the usage that counts against the limit at the given
// instant.
//
// A stored total naming a closed period reads as zero, because the row has not been
// cleared yet: clearing happens atomically with the first usage of the new period.
// Reporting the stale figure would show a token as exhausted through the whole
// first request of a fresh cycle, and the operator would see a quota that did not
// reset when it said it would.
//
// This mirrors DownstreamCredential.UsedTokensInCurrentPeriod, deliberately: the
// console and the admission path must not disagree about what has been spent.
func (r APITokenRow) UsedTokensNow(at time.Time) int64 {
	currentStamp := QuotaPeriodStamp(r.QuotaPeriod, at, LoadQuotaLocation(r.QuotaTimezone))
	if currentStamp == "" || r.QuotaPeriodStamp == currentStamp {
		return r.UsedTokens
	}
	// A different period *type* means the stamps are not comparable and the applying
	// statement carries the total forward, so it still counts.
	storedType, _, _ := strings.Cut(r.QuotaPeriodStamp, ":")
	currentType, _, _ := strings.Cut(currentStamp, ":")
	if storedType != currentType {
		return r.UsedTokens
	}
	return 0
}

// APITokenIn is the create payload. A nil Token means "generate one".
type APITokenIn struct {
	Name        string  `json:"name"`
	Description string  `json:"description"`
	Token       *string `json:"token"`
	Enabled     bool    `json:"enabled"`
	// ExpiresAt absent, null or blank means the token never expires.
	ExpiresAt *string `json:"expires_at"`
	// GroupID scopes which channels this token may reach. Absent means the
	// default group.
	GroupID *int64 `json:"group_id"`
	// LimitExpression is a token limit such as 100M or 1B. Blank means no limit.
	LimitExpression string `json:"limit_expression"`
	// RateLimit is a rate limit expression such as "100/m" or "1000/h". Blank means no limit.
	RateLimit *string `json:"rate_limit"`
	// AllowedModels restricts which models this credential may call. An absent
	// field, null and an empty array all mean unrestricted — the console can send
	// whichever its form produces without changing the outcome.
	AllowedModels []string `json:"allowed_models"`
	// QuotaPeriod is the reset cycle: none, daily, weekly or monthly. Absent or
	// blank means none, so a client that does not send the field gets the legacy
	// lifetime total.
	QuotaPeriod string `json:"quota_period"`
	// QuotaTimezone is the IANA zone the boundary falls in. Blank means UTC.
	QuotaTimezone string `json:"quota_timezone"`
}

// APITokenUpdateIn is a full replacement, so an absent `expires_at` clears the
// expiry rather than leaving it alone. The console always sends the field.
type APITokenUpdateIn struct {
	Name        string  `json:"name"`
	Description string  `json:"description"`
	ExpiresAt   *string `json:"expires_at"`
	GroupID     *int64  `json:"group_id"`
	// Token replaces the credential itself. Unlike ExpiresAt this is not a full
	// replacement: absent, null and blank all leave the current value alone. The
	// console echoes the token back on every save, and reading a blank field as
	// "erase it" would leave a row nobody can authenticate with.
	Token *string `json:"token"`
	// LimitExpression is a token limit such as 100M or 1B. Blank means no limit.
	LimitExpression string `json:"limit_expression"`
	// RateLimit is a rate limit expression such as "100/m" or "1000/h". Blank means no limit.
	RateLimit *string `json:"rate_limit"`
	// AllowedModels is a full replacement, like ExpiresAt and unlike Token: an
	// absent or empty array clears the restriction. That is what the console's
	// "clear the whitelist" action produces, and reading it as "leave it alone"
	// would make a restriction impossible to remove.
	AllowedModels []string `json:"allowed_models"`
	// QuotaPeriod and QuotaTimezone are full replacements too. An absent period
	// means none, which is how an operator turns automatic resets off.
	QuotaPeriod   string `json:"quota_period"`
	QuotaTimezone string `json:"quota_timezone"`
}

// RequestedToken is the replacement value, or "" when this edit keeps the
// current one.
func (t *APITokenUpdateIn) RequestedToken() string {
	if t.Token == nil {
		return ""
	}
	return *t.Token
}

// NormalizeRateLimit validates and trims a rate limit expression.
//
// A nil or blank value means "no rate limit" and is reported as nil, matching
// how the expiry field treats absence. The parsed form is discarded here — the
// stored shape is the expression itself, so the console can echo back exactly
// what the operator wrote.
func NormalizeRateLimit(raw *string) (*string, error) {
	if raw == nil {
		return nil, nil
	}
	value := strings.TrimSpace(*raw)
	if value == "" {
		return nil, nil
	}
	if _, err := ratelimit.ParseRateLimit(value); err != nil {
		return nil, ErrString("rate limit must look like 100/m, 1000/h or 50/10s")
	}
	return &value, nil
}

// validateTokenMetadata judges the name and description that will be stored, and
// writes the trimmed values back through the pointers it is given.
//
// The stores trim before writing, so the check has to be against the trimmed
// value or it answers about a different string than the one that lands in the
// database: a name padded past the limit with spaces was refused for a length it
// would not have had, and the emptiness check was already trimming while the
// length check beside it was not.
func validateTokenMetadata(name, description *string) error {
	trimmedName := strings.TrimSpace(*name)
	if trimmedName == "" || utf8.RuneCountInString(trimmedName) > APITokenNameMaxChars {
		return ErrString("token name must be between 1 and 80 characters")
	}
	if strings.ContainsFunc(trimmedName, unicode.IsControl) {
		return ErrString("token name must not contain control characters")
	}

	trimmedDescription := strings.TrimSpace(*description)
	if utf8.RuneCountInString(trimmedDescription) > APITokenDescriptionMaxChars {
		return ErrString("token description must be at most 200 characters")
	}
	if strings.ContainsFunc(trimmedDescription, unicode.IsControl) {
		return ErrString("token description must not contain control characters")
	}

	*name = trimmedName
	*description = trimmedDescription
	return nil
}

func isASCIIGraphic(value string) bool {
	for i := 0; i < len(value); i++ {
		if value[i] <= 0x20 || value[i] >= 0x7f {
			return false
		}
	}
	return true
}

// validateTokenValue judges an operator-supplied credential. Creation and
// editing share it so a value the create endpoint refuses cannot be smuggled in
// through an update.
func validateTokenValue(token string) error {
	if len(token) < APITokenMinBytes || len(token) > APITokenMaxBytes {
		return ErrString("custom token must be between 1 and 256 bytes")
	}
	if !isASCIIGraphic(token) {
		return ErrString("custom token must contain only printable ASCII characters without spaces")
	}
	return nil
}

func (t *APITokenIn) Validate() error {
	if err := validateTokenMetadata(&t.Name, &t.Description); err != nil {
		return err
	}
	if _, err := t.NormalizedExpiresAt(); err != nil {
		return err
	}
	if _, err := t.ParsedLimit(); err != nil {
		return err
	}
	if _, err := t.NormalizedRateLimit(); err != nil {
		return err
	}
	if _, err := t.NormalizedAllowedModels(); err != nil {
		return err
	}
	if _, _, err := t.NormalizedQuotaCycle(); err != nil {
		return err
	}
	if t.Token == nil {
		return nil
	}
	return validateTokenValue(*t.Token)
}

// NormalizedAllowedModels renders the whitelist for storage.
func (t *APITokenIn) NormalizedAllowedModels() (string, error) {
	return NormalizeAllowedModels(t.AllowedModels)
}

// NormalizedQuotaCycle validates and resolves the reset cycle for storage.
func (t *APITokenIn) NormalizedQuotaCycle() (period, timezone string, err error) {
	return normalizeQuotaCycle(t.QuotaPeriod, t.QuotaTimezone)
}

// NormalizeQuotaCycle is normalizeQuotaCycle for callers outside the console's
// create and update payloads — the configuration import writes a token row
// directly, and resolving the cycle a second way there is how the two would come
// to disagree about what a blank timezone means.
func NormalizeQuotaCycle(rawPeriod, rawTimezone string) (period, timezone string, err error) {
	return normalizeQuotaCycle(rawPeriod, rawTimezone)
}

// normalizeQuotaCycle resolves a submitted cycle, defaulting a blank period to
// none and a blank timezone to UTC.
//
// An unrecognised period is refused rather than silently defaulted: a console that
// sends "monthy" would otherwise store a token that never resets while its form
// shows a monthly cycle.
func normalizeQuotaCycle(rawPeriod, rawTimezone string) (string, string, error) {
	period := strings.ToLower(strings.TrimSpace(rawPeriod))
	if period == "" {
		period = DefaultQuotaPeriod
	}
	if !ValidQuotaPeriod(period) {
		return "", "", ErrString("quota_period must be none, daily, weekly or monthly")
	}

	timezone := strings.TrimSpace(rawTimezone)
	if timezone == "" {
		timezone = DefaultQuotaTimezone
	}
	if err := ValidateQuotaTimezone(timezone); err != nil {
		return "", "", err
	}
	return period, timezone, nil
}

func (t *APITokenIn) NormalizedExpiresAt() (*string, error) {
	return NormalizeExpiresAt(t.ExpiresAt)
}

// ParsedLimit resolves the limit expression into a stored token count.
func (t *APITokenIn) ParsedLimit() (*int64, error) {
	return ParseQuotaExpression(t.LimitExpression)
}

// NormalizedRateLimit validates the rate limit expression for storage.
func (t *APITokenIn) NormalizedRateLimit() (*string, error) {
	return NormalizeRateLimit(t.RateLimit)
}

func (t *APITokenUpdateIn) Validate() error {
	if err := validateTokenMetadata(&t.Name, &t.Description); err != nil {
		return err
	}
	if _, err := t.NormalizedExpiresAt(); err != nil {
		return err
	}
	if _, err := t.ParsedLimit(); err != nil {
		return err
	}
	if _, err := t.NormalizedRateLimit(); err != nil {
		return err
	}
	if _, err := t.NormalizedAllowedModels(); err != nil {
		return err
	}
	if _, _, err := t.NormalizedQuotaCycle(); err != nil {
		return err
	}
	// A blank token means "leave it alone", so it never reaches the value rules.
	if requested := t.RequestedToken(); requested != "" {
		return validateTokenValue(requested)
	}
	return nil
}

// NormalizedAllowedModels renders the whitelist for storage.
func (t *APITokenUpdateIn) NormalizedAllowedModels() (string, error) {
	return NormalizeAllowedModels(t.AllowedModels)
}

// NormalizedQuotaCycle validates and resolves the reset cycle for storage.
func (t *APITokenUpdateIn) NormalizedQuotaCycle() (period, timezone string, err error) {
	return normalizeQuotaCycle(t.QuotaPeriod, t.QuotaTimezone)
}

func (t *APITokenUpdateIn) NormalizedExpiresAt() (*string, error) {
	return NormalizeExpiresAt(t.ExpiresAt)
}

// ParsedLimit resolves the limit expression into a stored token count.
func (t *APITokenUpdateIn) ParsedLimit() (*int64, error) {
	return ParseQuotaExpression(t.LimitExpression)
}

// NormalizedRateLimit validates the rate limit expression for storage.
func (t *APITokenUpdateIn) NormalizedRateLimit() (*string, error) {
	return NormalizeRateLimit(t.RateLimit)
}

// APITokenOut carries the full token value so the console can hand a credential
// back to the operator who owns it.
type APITokenOut struct {
	ID          int64  `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description"`
	// Token is the plaintext, or "" for a row issued before plaintext was
	// stored — those cannot be recovered. Always serialized, never null, so a
	// client can test one field instead of distinguishing absent from empty.
	Token        string     `json:"token"`
	TokenPreview string     `json:"token_preview"`
	Enabled      bool       `json:"enabled"`
	ExpiresAt    *string    `json:"expires_at"`
	CreatedAt    string     `json:"created_at"`
	UpdatedAt    string     `json:"updated_at"`
	GroupID      int64      `json:"group_id"`
	GroupName    string     `json:"group_name"`
	Quota        QuotaState `json:"quota"`
	RateLimit    *string    `json:"rate_limit"`
	// AllowedModels is the operator's own list, always serialized as an array so
	// a client tests one field. An empty array is the wire form of "unrestricted";
	// there is no null variant, so no client has to distinguish the two.
	AllowedModels []string `json:"allowed_models"`
	// QuotaPeriodState reports the cycle, the current period and the next reset,
	// all derived server-side. The console must not compute a boundary itself: a
	// browser in another timezone would show a reset at the wrong hour, and one
	// with a skewed clock would show a period the gateway is not using.
	QuotaPeriodState QuotaPeriodState `json:"quota_period_state"`
}

// APITokenCreatedOut is what the creation endpoint answers with.
//
// It carries the full token, but so does APITokenOut: the console lets an
// operator copy a credential back out at any time, so token_plain is stored and
// the list and detail endpoints return it too. This is not a one-time reveal,
// and reading it as one understates where plaintext tokens are exposed.
type APITokenCreatedOut struct {
	ID               int64            `json:"id"`
	Name             string           `json:"name"`
	Description      string           `json:"description"`
	Token            string           `json:"token"`
	TokenPreview     string           `json:"token_preview"`
	Enabled          bool             `json:"enabled"`
	ExpiresAt        *string          `json:"expires_at"`
	CreatedAt        string           `json:"created_at"`
	UpdatedAt        string           `json:"updated_at"`
	GroupID          int64            `json:"group_id"`
	GroupName        string           `json:"group_name"`
	Quota            QuotaState       `json:"quota"`
	RateLimit        *string          `json:"rate_limit"`
	AllowedModels    []string         `json:"allowed_models"`
	QuotaPeriodState QuotaPeriodState `json:"quota_period_state"`
}

// TokenEnabledIn toggles a token.
//
// It mirrors UpstreamEnabledIn rather than reusing it, so the endpoint's
// contract names what it operates on, and it is a pointer for the same reason:
// a body of {} must not read as "disable this credential".
type TokenEnabledIn struct {
	Enabled *bool `json:"enabled"`
}

// Value returns the requested state, or an error when the body named none.
func (t *TokenEnabledIn) Value() (bool, error) {
	if t.Enabled == nil {
		return false, ErrString("enabled is required")
	}
	return *t.Enabled, nil
}
